relay-quotas/src/quota.rs

```diff
@@ -231,10 +232,71 @@ impl CategoryUnit {
 
 /// An efficient container for data categories that avoids allocations.
 ///
-/// [`DataCategories`] is a small-vector based collection of [`DataCategory`] values.
-/// It's optimized for the common case of having only a few categories, avoiding heap
-/// allocations in these scenarios.
-pub type DataCategories = SmallVec<[DataCategory; 8]>;
+/// It is a read only and has set like properties, allowing for fast comparisons.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Hash, Serialize)]
+#[serde(transparent)]
+pub struct DataCategories(Arc<[DataCategory]>);
+
+impl DataCategories {
+    /// Creates new and empty [`DataCategories`].
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Adds a data category to [`Self`].
+    ///
+    /// Returns `None` if the category was already contained, otherwise creates a new [`Self`] with
+    /// the `category` added.
+    pub fn add(&self, category: DataCategory) -> Option<Self> {
```

Comment:
Could this be 
```suggestion
    pub fn add(self, category: DataCategory) -> Self {
```

Or do we have cases where we don't have ownership / do not want to unnecessarily clone the `Arc`?

Reply:
Yeah that would work, I just liked giving the option to the caller.